### PR TITLE
optimize with defer logic

### PR DESCRIPTION
**Pull request by djunny**

### CODE REVIEW

File: src/functions.inc.php [hunk 1]
```diff
@@ -9,23 +9,17 @@
...
-
-            public function __construct(callable $callback)
-            {
-                $this->callback = $callback;
-            }
-
-            public function __destruct()
-            {
-                \call_user_func($this->callback);
+    $context = $context ?? new class extends SplStack {
```
Comment (jdreesen) [inline src/functions.inc.php]:
```suggestion
    $context ??= new class extends SplStack {
```

**Comment by bkrukowski** [comment]
What's the goal of this pull request?

**Comment by djunny** [comment]
less memory useage & higher performance & chain call & better implementation

**Comment by bkrukowski** [comment]
@djunny 

### Chain calls

Chain calls look a bit weird to me. I cannot see a big benefit of using

```php
defer($_, $callbackA)->push($callbackB);
```

instead of

```php
defer($_, $callbackA);
defer($_, $callbackB);
```

Furthermore, why do we have to register 2 callbacks in consecutive lines? Would not be easier to have just:

```php
defer($_, function() {
    $callbackB();
    $callbackA();
});
```

?

In my opinion, the above code is more explicit and easier to read. I appreciate your effort, but I have to reject this part.

### Design

I really like your way of making the order of callbacks explicit. 

```php
            while ($this->count() > 0) {
                \call_user_func($this->pop());
            }
```

If you are ok with reverting "chain calls" from your pull request, I would like to merge it.

**Comment by djunny** [comment]
chain calls:

I just return the SplStack object on the basis of the original, 

but i dont think it is recommended to use it like that.

so i removed in README

File: src/functions.inc.php [hunk 2]
```diff
@@ -9,23 +9,17 @@
  * file that was distributed with this source code.
  */
 
-function defer(?SplStack &$context, callable $callback): void
+function defer(?SplStack &$context, callable $callback): SplStack
```
Comment (bkrukowski) [inline src/functions.inc.php]: I cannot approve changing the header of this function, I explained it in the previous comment:

```suggestion
function defer(?SplStack &$context, callable $callback): void
```

File: README.md [hunk 3]
```diff
@@ -27,14 +27,20 @@ defer($context, $callback);
...
+    // 7.3 -
     defer($_, function () {
         echo "goodbye\n";
     });
 
     defer($_, function () {
         echo "...\n";
     });
-
+
```
Comment (bkrukowski) [inline README.md]: Please revert those changes

**Review by bkrukowski — APPROVED** [review]

**Pull request merged**